apps/kg/models/pytorch/score_fun.py

```diff
@@ -333,3 +333,97 @@ def fn(heads, relations, tails, num_chunks, chunk_size, neg_sample_size):
                 tmp = tmp.reshape(num_chunks, chunk_size, hidden_dim)
                 return th.bmm(tmp, tails)
             return fn
+
+class RotatEScore(nn.Module):
+    def __init__(self, gamma, emb_init):
+        super(RotatEScore, self).__init__()
+        self.gamma = gamma
+        self.emb_init = emb_init
+
+    def edge_func(self, edges):
+        pi = 3.14159265358979323846
+        re_head, im_head = th.chunk(edges.src['emb'], 2, dim=-1)
+        re_tail, im_tail = th.chunk(edges.dst['emb'], 2, dim=-1)
```

Comment:
Somthings are little different when dealing with complex numbers. I think if we don't split tail, the code would be like this
```
score = th.cat([re_score, im_score], dim=-1)-tail
re_score, im_score =th.chunk(score, 2, dim=-1)
score = th.stack([re_score, im_score], dim=0)
score = score.norm(-1)
return gamma - score.sum(-1)
```
which seems more redundant to me. I guess that your initial idea was 
```
score = th.cat([re_score, im_score], dim=-1)-tail
score = score.norm(-1)
return gamma - score.sum(-1)
```
which is incorrect since the formula should be
$$ \sqrt{s_{re_1}^2+s_{im_1}^2}+...+\sqrt{s_{re_d}^2+s_{im_d}^2}.$$
If we use the second code above, what we are actually calculating
$$ \sqrt{s_{re_1}^2+s_{im_1}^2+...+s_{re_d}^2+s_{im_d}^2}.$$
They are not equivalant to each other. I was actually stuck on this issue for  long time debugging till I found out the reason behind.

Reply:
I see. Actually there are two dimensions here.